server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -226,22 +238,45 @@ class AsyncAction {
         private final NodesRequest request;
         private final ActionListener<NodesResponse> listener;
         private final AtomicReferenceArray<Object> responses;
+        private final DiscoveryNode[] concreteNodes;
         private final AtomicInteger counter = new AtomicInteger();
         private final Task task;
 
         AsyncAction(Task task, NodesRequest request, ActionListener<NodesResponse> listener) {
             this.task = task;
             this.request = request;
             this.listener = listener;
-            if (request.concreteNodes() == null) {
+
+            // Check if concrete nodes are already available
+            if (request.concreteNodes() != null) {
+                this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
+
+                if (request.populateDiscoveryNodesInTransportRequest()) {
+                    this.concreteNodes = null;
+                } else {
+                    this.concreteNodes = request.concreteNodes();
+                    request.setConcreteNodes(null);
+                    assert request.concreteNodes() == null;
+                }
+                return;
+            }
+
+            // Check if we want to populate the DiscoveryNodes in the transport Request and accordingly backfill the
+            // concrete nodes.
+            if (request.populateDiscoveryNodesInTransportRequest()) {
                 resolveRequest(request, clusterService.state());
                 assert request.concreteNodes() != null;
+                this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
+                this.concreteNodes = null;
+            } else {
+                this.concreteNodes = getConcreteNodes(request, clusterService.state());
+                assert request.concreteNodes() == null;
+                this.responses = new AtomicReferenceArray<>(concreteNodes.length);
```

Comment:
The code looks confusing now.

You only need simple logic whether to serialize request.concreteNodes or not. You can request.concreteNodes to false based on the flag. Everything else can remain the same.

Reply:
The reason why I am explicitly removing concreteNodes entirely from the request as well is to remove the amount of ConcreteNodes set in JVM memory at the time of the request being made. 

While the serialisation and deserialisation latencies will improve by taking this decision during the IO, the overall pressure on the coordinator node to store these number of heavy requests will remain the same. Also the request payload itself too will remain the same.